php-frontend/src/main/java/org/sonar/php/parser/TreeFactory.java

```diff
@@ -0,0 +1,1100 @@
+      InternalSyntaxToken closeParenthesisToken
+  ) {
+    return new ForEachStatementHeader(
+        forEachToken, openParenthesisToken,
+        expression, asToken, getForEachKey(optionalKey), getForEachArrow(optionalKey), value,
+        closeParenthesisToken
+    );
+  }
+
+  @Nullable
+  private ExpressionTree getForEachKey(Optional<Tuple<ExpressionTree, InternalSyntaxToken>> optionalKey) {
+    if (optionalKey.isPresent()) {
+      return optionalKey.get().first();
+    } else {
+      return null;
+    }
+  }
+
+  @Nullable
+  private InternalSyntaxToken getForEachArrow(Optional<Tuple<ExpressionTree, InternalSyntaxToken>> optionalKey) {
+    if (optionalKey.isPresent()) {
+      return optionalKey.get().second();
+    } else {
+      return null;
+    }
+  }
+
+  public ForStatementHeader forStatementHeader(
+      InternalSyntaxToken forToken, InternalSyntaxToken lParenthesis,
+      Optional<SeparatedList<ExpressionTree>> init, InternalSyntaxToken semicolon1,
+      Optional<SeparatedList<ExpressionTree>> condition, InternalSyntaxToken semicolon2,
+      Optional<SeparatedList<ExpressionTree>> update, InternalSyntaxToken rParenthesis
+  ) {
+    return new ForStatementHeader(
+        forToken, lParenthesis,
+        optionalSeparatedList(init),
+        semicolon1,
+        optionalSeparatedList(condition),
+        semicolon2,
+        optionalSeparatedList(update),
+        rParenthesis
+    );
+  }
+
+  public ForStatementTree forStatement(ForStatementHeader forStatementHeader, StatementTree statement) {
+    return new ForStatementTreeImpl(forStatementHeader, statement);
+  }
+
+  public ForStatementTree forStatementAlternative(
+      ForStatementHeader forStatementHeader, InternalSyntaxToken colonToken,
+      Optional<List<StatementTree>> statements, InternalSyntaxToken endForToken, InternalSyntaxToken eos
+  ) {
+    return new ForStatementTreeImpl(forStatementHeader, colonToken, optionalList(statements), endForToken, eos);
+  }
+
+  public SeparatedList<ExpressionTree> forExpr(ExpressionTree expression, Optional<List<Tuple<InternalSyntaxToken, ExpressionTree>>> listOptional) {
+    return separatedList(expression, listOptional);
+  }
+
+  public ElseClauseTree elseClause(InternalSyntaxToken elseToken, StatementTree statement) {
+    return new ElseClauseTreeImpl(elseToken, statement);
+  }
+
+  public IfStatementTree ifStatement(
+      InternalSyntaxToken ifToken, ParenthesisedExpressionTree expression, StatementTree statement,
+      Optional<List<ElseifClauseTree>> elseIfClauses, Optional<ElseClauseTree> elseClause
+  ) {
+    return new IfStatementTreeImpl(ifToken, expression, statement, optionalList(elseIfClauses), elseClause.orNull());
+  }
+
+  public ElseifClauseTree elseifClause(InternalSyntaxToken elseifToken, ParenthesisedExpressionTree condition, StatementTree statement) {
+    return new ElseifClauseTreeImpl(elseifToken, condition, statement);
+  }
+
+  public IfStatementTree alternativeIfStatement(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)